extensions/amp-accordion/1.0/accordion.js

```diff
@@ -0,0 +1,176 @@
+  /** @type {AccordionDef.ContextProps} */ ({})
+);
+
+/** @type {!Object<string, boolean>} */
+const EMPTY_EXPANDED_SET = {};
+
+const generateSectionId = sequentialIdGenerator();
+
+/**
+ * @param {!AccordionDef.Props} props
+ * @return {PreactDef.Renderable}
+ */
+export function Accordion({
+  as: Comp = 'section',
+  expandSingleSection = false,
+  children,
+  ...rest
+}) {
+  const [expandedSet, setExpandedSet] = useState(EMPTY_EXPANDED_SET);
+
+  useEffect(() => {
+    if (!expandSingleSection) {
+      return;
+    }
+    setExpandedSet((expandedSet) => {
+      const newExpandedSet = {};
+      let expanded = 0;
+      for (const k in expandedSet) {
+        newExpandedSet[k] = expandedSet[k] && expanded++ == 0;
+      }
+      return newExpandedSet;
+    });
+  }, [expandSingleSection]);
+
+  const registerSection = useCallback((id) => {
+    return () => setExpandedSet((expandedSet) => omit(expandedSet, id));
+  }, []);
+
+  const toggleExpanded = useCallback(
+    (id, defaultExpanded) => {
+      setExpandedSet((expandedSet) => {
+        const newValue = !isExpanded(
+          id,
+          defaultExpanded,
+          expandedSet,
+          expandSingleSection
+        );
+        let newExpandedSet;
+        if (newValue && expandSingleSection) {
+          newExpandedSet = {[id]: newValue};
+          for (const k in expandedSet) {
+            if (k != id) {
+              newExpandedSet[k] = false;
+            }
+          }
+        } else {
+          newExpandedSet = {...expandedSet, [id]: newValue};
+        }
+        return newExpandedSet;
+      });
+    },
+    [expandSingleSection]
+  );
+
+  const context = useMemo(
+    () =>
+      /** @type {!AccordionDef.ContextProps} */ ({
+        registerSection,
+        toggleExpanded,
+        isExpanded: (id, defaultExpanded) =>
+          isExpanded(id, defaultExpanded, expandedSet, expandSingleSection),
+      }),
+    [expandedSet, expandSingleSection, registerSection, toggleExpanded]
+  );
+
+  return (
+    <Comp {...rest}>
+      <AccordionContext.Provider value={context}>
+        {children}
+      </AccordionContext.Provider>
+    </Comp>
+  );
+}
+
+/**
+ * @param {string} id
+ * @param {boolean} defaultExpanded
+ * @param {!Object<string, boolean>} expandedSet
+ * @param {boolean} expandSingleSection
+ * @return {boolean}
+ */
+function isExpanded(id, defaultExpanded, expandedSet, expandSingleSection) {
+  const current = expandedSet[id];
+  if (current != null) {
+    return current;
+  }
+  const adjDefaultExpanded = expandSingleSection
+    ? Object.values(expandedSet).includes(true)
+      ? false
+      : defaultExpanded
+    : defaultExpanded;
+  return adjDefaultExpanded;
+}
+
+/**
+ * @param {!AccordionDef.SectionProps} props
+ * @return {PreactDef.Renderable}
+ */
+export function AccordionSection({
+  as: Comp = 'section',
+  headerAs: HeaderComp = 'header',
+  contentAs: ContentComp = 'div',
+  expanded: defaultExpanded = false,
```

Comment:
Can the `expanded` prop not be mutated on the section to change its state? Looks like the accordion state takes precedence?

Reply:
`expanded` prop does not need to be mutated in this model. I.e. the prop value is only used as a default the first time.